be/src/exec/vectorized/partition/chunks_partitioner.cpp

```diff
@@ -0,0 +1,232 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#include "exec/vectorized/partition/chunks_partitioner.h"
+
+#include "exprs/expr.h"
+#include "exprs/expr_context.h"
+#include "runtime/current_thread.h"
+
+namespace starrocks::vectorized {
+
+// TODO(hcf) move to utils
+#define RETURN_PTYPE_BYTE_SIZE(TYPE, SIZE)                                           \
+    case TYPE:                                                                       \
+        static_assert(sizeof(vectorized::RunTimeTypeTraits<TYPE>::CppType) == SIZE); \
+        return SIZE;
+
+inline static int get_byte_size_of_primitive_type(PrimitiveType type) {
+    switch (type) {
+        RETURN_PTYPE_BYTE_SIZE(TYPE_NULL, 1);
+        RETURN_PTYPE_BYTE_SIZE(TYPE_BOOLEAN, 1);
+        RETURN_PTYPE_BYTE_SIZE(TYPE_TINYINT, 1);
+
+        RETURN_PTYPE_BYTE_SIZE(TYPE_SMALLINT, 2);
+
+        RETURN_PTYPE_BYTE_SIZE(TYPE_DECIMAL32, 4);
+        RETURN_PTYPE_BYTE_SIZE(TYPE_DATE, 4);
+        RETURN_PTYPE_BYTE_SIZE(TYPE_INT, 4);
+        RETURN_PTYPE_BYTE_SIZE(TYPE_FLOAT, 4);
+
+        RETURN_PTYPE_BYTE_SIZE(TYPE_DECIMAL64, 8);
+        RETURN_PTYPE_BYTE_SIZE(TYPE_BIGINT, 8);
+        RETURN_PTYPE_BYTE_SIZE(TYPE_TIME, 8);
+        RETURN_PTYPE_BYTE_SIZE(TYPE_DATETIME, 8);
+        RETURN_PTYPE_BYTE_SIZE(TYPE_DOUBLE, 8);
+
+        RETURN_PTYPE_BYTE_SIZE(TYPE_DECIMAL128, 16);
+        RETURN_PTYPE_BYTE_SIZE(TYPE_LARGEINT, 16);
+        RETURN_PTYPE_BYTE_SIZE(TYPE_DECIMALV2, 16);
+    default:
+        return 0;
+    }
+}
+
+ChunksPartitioner::ChunksPartitioner(const bool has_nullable_partition_column,
+                                     const std::vector<ExprContext*>& partition_exprs,
+                                     const std::vector<PartitionColumnType>& partition_types)
+        : _has_nullable_partition_column(has_nullable_partition_column),
+          _partition_exprs(partition_exprs),
+          _partition_types(partition_types) {
+    _partition_columns.resize(partition_exprs.size());
+}
+
+Status ChunksPartitioner::prepare(RuntimeState* state) {
+    _state = state;
+    _init_hash_map_variant();
+    return Status::OK();
+}
+
+Status ChunksPartitioner::offer(const ChunkPtr& chunk) {
+    DCHECK(!_partition_it.has_value());
+
+    for (size_t i = 0; i < _partition_exprs.size(); i++) {
+        ASSIGN_OR_RETURN(_partition_columns[i], _partition_exprs[i]->evaluate(chunk.get()));
+    }
+
+    if (false) {
```

Comment:
What's the meaning of this statement?

Reply:
the macro below will generate bunch of `else if` statement, so for the integrity of grammar, we need a header `if(false)` statement

Reply:
Why not use a switch-case?